service/lib/dinstaller/storage/manager.rb

```diff
@@ -20,41 +20,73 @@
 # find current contact information at www.suse.com.
 
 require "yast"
+require "bootloader/proposal_client"
+require "bootloader/finish_client"
 require "y2storage/storage_manager"
 require "dinstaller/storage/proposal"
 require "dinstaller/storage/callbacks"
+require "dinstaller/with_progress"
+require "dinstaller/can_ask_question"
+require "dinstaller/security"
+require "dinstaller/dbus/clients/questions_manager"
+require "dinstaller/helpers"
 
 Yast.import "PackagesProposal"
 
 module DInstaller
   module Storage
     # Manager to handle storage configuration
     class Manager
-      def initialize(logger, config)
-        @logger = logger
+      include WithProgress
+      include CanAskQuestion
+      include Helpers
+
+      def initialize(config, logger)
         @config = config
+        @logger = logger
       end
 
       # Probes storage devices and performs an initial proposal
-      #
-      # @param questions_manager [QuestionsManager]
-      def probe(questions_manager)
-        # TODO: Add progress once this is moved to its own service
-        logger.info "Probing storage and performing proposal"
-        activate_devices(questions_manager)
-        probe_devices
-        proposal.calculate
+      def probe
+        start_progress(4)
+        progress.step("Activating storage devices") { activate_devices }
+        progress.step("Probing storage devices") { probe_devices }
+        progress.step("Calculating the storage proposal") { proposal.calculate }
+        progress.step("Selecting Linux Security Modules") { security.probe }
       end
 
       # Prepares the partitioning to install the system
       def install
-        add_packages
-        Yast::WFM.CallFunction("inst_prepdisk", [])
+        start_progress(4)
+        progress.step("Preparing bootloader proposal") do
+          # first make bootloader proposal to be sure that required packages are installed
```

Comment:
hmm, how this will work with separate processes? it first says which packages bootloader needs and then writes bootloader, but where it wait till it is installed?

Reply:
At that point, it adds the packages to the PackagesProposal (through D-Bus, thanks to our [overriden PackagesProposal](https://github.com/yast/d-installer/blob/master/service/lib/dinstaller/dbus/y2dir/manager/modules/PackagesProposal.rb).

Later, when the [bootloader is about to be installed](https://github.com/yast/d-installer/pull/248/files/5c7d9beeb80892a958e2138c0904250cb8a46532#diff-808c5e766d91d2840280e4f4ee5248f8a434ab7d78cd6e9a21103cc317e8eb72R84), the packages should be there.